net/lp_rpc.proto

```diff
@@ -20,12 +20,62 @@ message TranscoderRequest {
   bytes sig   = 2;
 }
 
+/*
+  OSInfo needed to negotiate storages that will be used.
+  It carries info needed to write to the storage.
+*/
+message OSInfo {
+
+  // Storage type: s3, ipfs.
+  string storage = 1;
+
+  S3OSInfo s3info = 16;
```

Comment:
it can, but IPFS driver will not use fields that not intended to it, so it's not a problem.
plus it is `protobuf` - what in protobuf's type system can be used to constrain states?

Reply:
Mostly thinking about `oneof` even though it's a little bit of a pain to use in golang. See https://github.com/livepeer/go-livepeer/issues/582 Not really important either way.

Reply:
We still need to use `storage` (be it string or enum) to define type, `oneof` will server same purpose, but only for one special - so we will be using two different way to define type. not good.

Reply:
Not sure if I understand. Oneof eliminates the redundancy on the wire.  It also eliminates a few classes of data consistency errors by constraining what's representable. For example:

1. Error of conflicting type: IPFS storage type set, but have S3 also configured
2. Error of omission: Configure S3, but have a missing storage type string

Note that you need to check the type no matter which method we use, but you have stronger assurances of well-formed data with oneof.

```golang
message OSInfo {
  oneof storageType {
    S3Info s3
    string otherStorage
  }
}

switch msg.StorageType.(type) {
  case *net.OsInfo_S3:
    // initialize S3 driver
  case *net.OSInfo_OtherStorage:
    // May be "ipfs" or "direct" . "s3" is not valid here as it's missing the S3 data.
    initialize some other driver
}
```

Note that we may also elect to have individual `bool` fields for each unary storage type (rather than a string or enum) but that's perhaps more brittle.

Reply:
here
```
 initialize some other driver
``` 
you will have another switch:
```golang
switch msg.StorageType.(type) {
  case *net.OsInfo_S3:
    // initialize S3 driver
  case *net.OSInfo_OtherStorage:
    // May be "ipfs" or "direct" . "s3" is not valid here as it's missing the S3 data.
    switch *net.OSInfo_OtherStorage {
       case "ipfs: initialzie ipfs
       case "azure": initialize azure
    }
}
```

so two switches, both basically doing same thing - checking type in runtime

Reply:
That's why I mentioned the possibility of individual bool fields for specific storage values, then you'd only need one switch.

In any case -- the additional switch here is stylistic. What's more important IMO is promoting data consistency and reducing redundancy on the wire.